x-template.js

```diff
@@ -0,0 +1,1062 @@
+  }
+
+  static #insertAllBefore(parentNode, referenceNode, nodes) {
+    // Iterate backwards over the live node collection since we’re mutating it.
+    // Note that passing “null” as the reference node appends nodes to the end.
+    for (let iii = nodes.length - 1; iii >= 0; iii--) {
+      const node = nodes[iii];
+      parentNode.insertBefore(node, referenceNode);
+      referenceNode = node;
+    }
+  }
+
+  static #removeWithin(node) {
+    // Iterate backwards over the live node collection since we’re mutating it.
+    const childNodes = node.childNodes;
+    for (let iii = childNodes.length - 1; iii >= 0; iii--) {
+      node.removeChild(childNodes[iii]);
+    }
+  }
+
+  static #removeBetween(startNode, node) {
+    while(node.previousSibling !== startNode) {
+      node.previousSibling.remove();
+    }
+  }
+
+  static #removeThrough(startNode, node) {
+    TemplateEngine.#removeBetween(startNode, node);
+    startNode.remove();
+    node.remove();
+  }
+
+  static #clearObject(object) {
+    for (const key of Object.keys(object)) {
+      delete object[key];
+    }
+  }
+
+  // TODO: Replace with Map.prototype.getOrInsert when TC39 proposal lands.
+  //  https://github.com/tc39/proposal-upsert
+  static #setIfMissing(map, key, callback) {
+    // Values set in this file are ALL truthy, so "get" is used (versus "has").
+    let value = map.get(key);
+    if (!value) {
+      value = callback();
+      map.set(key, value);
+    }
+    return value;
+  }
+
+  static #getState(object, key) {
+    // Values set in this file are ALL truthy.
+    let value = object[key];
+    if (!value) {
+      value = {};
+      object[key] = value;
+    }
+    return value;
+  }
+
+  static #getBindingText(binding) {
+    switch (binding) {
+      case TemplateEngine.#ATTRIBUTE: return 'an attribute';
+      case TemplateEngine.#BOOLEAN: return 'a boolean attribute';
+      case TemplateEngine.#DEFINED: return 'a defined attribute';
+      case TemplateEngine.#PROPERTY: return 'a property';
+      case TemplateEngine.#CONTENT: return 'content';
+      case TemplateEngine.#TEXT: return 'text content';
+    }
+  }
+
+  static #interfaceDeprecatedStacks = new Set();
+  static #interfaceDeprecated(name, callback) {
+    return (...args) => {
+      const error = new Error(`Deprecated "${name}" from default templating engine interface.`);
+      const stack = error.stack;
+      if (!this.#interfaceDeprecatedStacks.has(stack)) {
+        this.#interfaceDeprecatedStacks.add(stack);
+        console.warn(error); // eslint-disable-line no-console
+      }
+      return callback(...args);
+    };
+  }
+
+  static #interfaceRemoved(name) {
+    return () => {
+      throw new Error(`Removed "${name}" from default templating engine interface. Import and plug-in "lit-html" as your element's templating engine if you want this functionality.`);
+    };
+  }
+}
+
+// Long-term interface.
+export const render = TemplateEngine.interface.render.bind(TemplateEngine);
```

Comment:
Rather than export the class (which would lend itself to potential extension / misuse)… we export the exact functions in the interface (many of which are soon to be deleted!).